numpy/core/tests/test_nditer.py

```diff
@@ -2871,6 +2873,13 @@ def add_context(x, y, out=None):
     z = add_context(range(5), range(5))
     assert_equal(z, range(0, 10, 2))
 
+def test_close_raises():
+    it = np.nditer(np.arange(3))
+    assert_equal (next(it), 0)
+    it.close()
+    assert_raises(StopIteration, next, it)
+    assert_raises(ValueError, getattr, it, 'operands')
```

Comment:
I tried to document the behavour wrt writeback semantics [here](https://github.com/numpy/numpy/blob/master/numpy/add_newdocs.py#L401)
If writeback semantics were used and have been resolved, the operand's data is no longer related to the input array's data. If writeback semantics were not used, the operand shares data with the input array. So we could create a situation where code is fragile: it works without writeback semantics but fails when they are used.

Note this was true even before PR #9998. The previous trigger to resolve was a call to `NpyIter_Deallocate`, which was done by `del it` or `it = None`. Since `it` was no longer valid, new access to the operands was prevented.

We started discussing this in the nditer.close pull request in [Dec](https://github.com/numpy/numpy/pull/10184#pullrequestreview-83745185), but I guess it never really got resolved.

I am open to suggestions. The motivation/root-cause of this whole mess is that the operand can have a different, non-compatible dtype from the array. At some point that incompatibility needs to be resolved, and from that point forward the operand is unusable or we need some kind of flag to track between the (now writeable) data in the array and the still-valid(?), writeable operand.

Reply:
OK, you're right, I was misunderstanding it. The `it.operands` are indeed the temporary buffers, and not the final output array, and so should not be accessed after closing. Eg,
```python
>>> a = np.arange(24, dtype='f8')
>>> aiter = np.nditer(a, [], [['readwrite', 'updateifcopy']],
...                  casting='same_kind',op_dtypes=[np.dtype('f4')])
>>> with aiter as it:
...     x = it.operands[0]
>>> x[0] = 10
>>> a[0]
0.0
```

But in that case all the examples above which do things like `return it.operands[1]` don't make sense, since they appear to be returning the temp buffer rather than the final array. We need to provide a way for the user to access the allocated output arrays even after the iterator is closed.

Perhaps `it.operators` should point to the actual array instead of the temp buffer? Or (more likely) maybe we need to add a new attribute to the nditer object which points to the final arrays.

Reply:
The nditer docstring says:
```
operands | (tuple of operand(s)) The array(s) to be iterated over.
```
which sounds to me like it is describing the "original" arrays, not the buffer copies. All of the code which does `return it.operands[1]` seems to implicitly assume `it.operands` refers to the original arrays.

The one exception in the examples is in reductions, which start by doing `it.operands[1][..] = 0` to initialize the reduction, which assumes that `it.operands` refers to the buffers. However, there seems to be a whole complicated discussion of a special flag `delay_bufalloc` and a special method `it.reset()` which, looking at it in retrospect, seem to me to be hacks only necessary because of the lack of distinction between the "original" arrays and the buffers which we have discovered here. 

Maybe now that we have a context manager, we can deprecate all of that and add some better behavior? That is, you would do `it.operands[1][..] = 0` before the context is entered, (so context entrance replaces `it.reset`)?

 Maybe that's all going too far, though.